baseplate/queue_consumer.py

```diff
@@ -0,0 +1,127 @@
+    exits. If there is an error in processing and the message must be retried
+    the handler should raise an exception to crash the process. This will
+    prevent the ack and the message will be requeued at the head of the queue.
+
+    :param baseplate.core.Baseplate baseplate: A baseplate instance for the
+        service.
+    :param kombu.Exchange exchange:
+    :param kombu.connection.Connection connection:
+    :param str queue_name: The name of the queue.
+    :param list routing_keys: List of routing keys.
+    :param handler: The handler method.
+
+    """
+    queues = []
+    for routing_key in routing_keys:
+        queue = Queue(
+            name=queue_name,
+            exchange=exchange,
+            routing_key=routing_key,
+        )
+        queues.append(queue)
+
+    print("registering %s as a handler for %r" % (handler.__name__, queues))
+    kombu_consumer = KombuConsumer(baseplate, connection, exchange, queues)
+    kombu_consumer.start()
+
+    print("waiting for messages")
+    while True:
+        with kombu_consumer.get_message() as (context, message, span):
+            handler(context, message.body, message)
+            message.ack()
+
+
+class _ConsumerWorker(ConsumerMixin):
+    def __init__(self, connection, queues, work_queue):
+        self.connection = connection
+        self.queues = queues
+        self.work_queue = work_queue
+
+    def get_consumers(self, Consumer, channel):
+        return [Consumer(
+            queues=self.queues,
+            on_message=self.on_message,
+        )]
+
+    def on_message(self, message):
+        self.work_queue.put(message)
+
+
+class ConsumerContext(object):
+    pass
+
+
+class KombuConsumer(object):
+    def __init__(self, baseplate, connection, exchange, queues):
+        self.baseplate = baseplate
+        self.connection = connection
+        self.exchange = exchange
+        self.queues = queues
+        self.work_queue = queue.Queue()
+        self.consumer = _ConsumerWorker(connection, queues, self.work_queue)
+        self.thread = None
+
+    def start(self):
+        assert self.thread is None, "consumer already started"
+        self.thread = Thread(target=self.consumer.run)
+        self.thread.name = "consumer message pump"
+        self.thread.daemon = True
+        self.thread.start()
+
+    @contextlib.contextmanager
+    def get_message(self, timeout=None):
+        with self.get_batch(max_items=1, timeout=timeout) as (context, batch, span):
+            yield context, batch[0], span
+
+    @contextlib.contextmanager
+    def get_batch(self, max_items, timeout):
+        assert self.thread is not None, "you must start the consumer first"
+
+        if timeout == 0:
+            block = False
+        else:
+            block = True
+
+        batch = []
+        for _ in range(max_items):
+            try:
+                item = self.work_queue.get(block=block, timeout=timeout)
```

Comment:
should `timeout` be a "per check" timeout? or maximum timeout? i.e. should it be decreasing as time progresses across multiple `work_queue.get()` calls?

Reply:
I'm not sure. In the original version of this that I copied from you there was a TODO comment about this. What is the thinking on using a variable timeout?

Reply:
If the timeout=1 and max_items is 3 we could end up waiting 3 seconds (1 second per item) for a full batch when you might expect that you'd time out at 1 total second regardless.

It just occurred to me that we can use baseplate's retry stuff for this:

```python
for time_remaining in RetryPolicy.new(attempts=max_items, budget=timeout):
    try:
        item = self.work_queue.get(block=block, timeout=time_remaining)
        batch.append(item)
    except queue.Empty:
        break
```